src/petals/server/memory_cache.py

```diff
@@ -172,5 +204,32 @@ def use_cache(self, *handles: Handle) -> Sequence[torch.Tensor]:
         yield tuple(self._allocated_tensors[handle] for handle in handles)
 
 
+@contextlib.asynccontextmanager
+async def wait_for_aenter(context: contextlib.AbstractAsyncContextManager, stop_when_completes: Coroutine):
+    """Try to enter asynchronous context in time before stop_after task completes (e.g. timeout)"""
+
+    async def _enter_and_exit():
+        async with context:
+            yield
+        yield
+
+    async def _wait_for_deadline():
+        await stop_when_completes
+
+    aenter_and_aexit = _enter_and_exit()
+    aenter_task = asyncio.create_task(anext(aenter_and_aexit))
+    stop_task = asyncio.create_task(_wait_for_deadline())
+    try:
+        await asyncio.wait({aenter_task, stop_task}, return_when=asyncio.FIRST_COMPLETED)
+        if stop_task.done():
+            raise TimeoutError("Did not enter context in time")
+        yield aenter_task.result()
+    finally:
+        if aenter_task.done():
+            await anext(aenter_and_aexit)  # exit normally
```

Comment:
**Important:** `await` in finally does not always work (see above comments and #396). Can we simplify this function without it?

Reply:
it's gone, good catch!